File: 088_Merge_Sorted_Array/088_1.py
from typing import List, Dict

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        def binary_search(target, lo, hi):
            while lo < hi:
                mid = (lo + hi) >> 1
                if target <= nums1[mid]:
                    hi = mid
                elif target > nums1[mid]:
                    lo = mid + 1
            return lo

        k = 0
        for i in range(n):
            k = binary_search(nums2[i], k, m)
            for j in range(m, k, -1):
                nums1[j] = nums1[j-1]
            nums1[k] = nums2[i]
            m += 1


if __name__ == "__main__":
    nums1 = [1,2,3,0,0,0]
    nums2 = [2,5,6]
    Solution().merge(nums1, 3, nums2, 3)
    print(nums1)

File: 088_Merge_Sorted_Array/088_1.c
void merge(int* nums1, int nums1Size, int m, int* nums2, int nums2Size, int n){
    int k = m + n - 1;
    m--;
    n--;
    while (m >= 0 && n >= 0) {
        nums1[k--] = nums2[n] > nums1[m] ? nums2[n--] : nums1[m--];
    }
   // while (n >= 0) {
   //     nums1[k--] = nums2[n--];
   // }
    if (n >= 0)
        memcpy(nums1, nums2, (n+1) * sizeof(int));
}
